emailproxy.py

```diff
@@ -337,9 +341,141 @@ def add_account(username):
+                            try:
+                                aws_client.create_secret(
+                                    Name=secret_id,
+                                    ForceOverwriteReplicaSecret=False
+                                )
+                            except botocore.exceptions.ClientError as err_createsecret:
+                                if err_createsecret.response['Error']['Code'] == 'AccessDeniedException':
+                                    Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s" - access denied in attempt to create it. does your IAM user have "secretsmanager:CreateSecret" permissions?'  % (secret_id))
+                                else:
+                                    Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s" - attempt to create it failed with unexpected error, see debug logs for details')
+                                AppConfig.aws_secrets_debuglog_boto3error(err_createsecret)
+                            else:
+                                Log.info('Created AWS Secret "%s"' % (secret_id))
+                                try:
+                                    aws_client.put_secret_value(
+                                        SecretId=secret_id,
+                                        SecretString=json.dumps(aws_secrets[secret_id]),
+                                    )
+                                except botocore.exceptions.ClientError as err_putsecret_aftercreate:
+                                    if err_putsecret_aftercreate.response['Error']['Code'] == 'AccessDeniedException':
+                                        Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s" - access denied, does IAM user have "secretsmanager:PutSecretValue" permissions?' % (secret_id))
+                                    else:
+                                        Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s" - unexpected error, see debug logs for details')
+                                    AppConfig.aws_secrets_debuglog_boto3error(err_putsecret_aftercreate)
+                    elif err_putsecret.response['Error']['Code'] == 'AccessDeniedException':
+                        Log.error('Failed to store OAuth tokens in AWS Secret "%s" - access denied, does IAM user have "secretsmanager:PutSecretValue" permissions?' % (secret_id))
+                        AppConfig.aws_secrets_debuglog_boto3error(err_putsecret)
+                    else:
+                        Log.error('Failed to store OAuth tokens in AWS Secret "%s" - unexpected error, see debug logs for details' % (secret_id))
+                        AppConfig.aws_secrets_debuglog_boto3error(err_putsecret)
+        else:
+            Log.error('Failed to store OAuth 2.0 tokens in AWS Secrets Manager')
 
+        # Return copy of config without OAuth 2.0 tokens to write to disk
+        return appconfig_to_save
```

Comment:
```suggestion
    def _aws_secrets_save():
        # a deep copy of the configuration file to be saved locally, *without* OAuth 2.0 tokens (_SECRETS_MANAGER_KEYS)
        appconfig_to_save = configparser.ConfigParser()
        appconfig_to_save.read_dict(AppConfig._PARSER)

        # remove OAuth 2.0 tokens from each account, transferring them into a local dict of referenced AWS Secrets
        aws_secrets = AppConfig._aws_secrets_ids()
        for account in AppConfig._aws_secrets_accounts():
            account_secret = appconfig_to_save.get(account, 'aws_secret')
            aws_secrets[account_secret][account] = {key: appconfig_to_save.get(account, key) for key in
                                                    AppConfig._SECRETS_MANAGER_KEYS}
            for key in AppConfig._SECRETS_MANAGER_KEYS:
                appconfig_to_save.remove_option(account, key)

        # save the extracted AWS Secrets, attempting to create new secrets if they do not already exist
        aws_client = AppConfig._aws_secrets_boto3_client()
        if aws_client:
            for secret_id in aws_secrets:
                try:
                    aws_client.put_secret_value(SecretId=secret_id, SecretString=json.dumps(aws_secrets[secret_id]))

                except botocore.exceptions.ClientError as err_put_secret:
                    if err_put_secret.response['Error']['Code'] == 'ResourceNotFoundException':
                        if secret_id.startswith('arn:'):
                            Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s"' % secret_id, '- the Secret',
                                      'does not exist, and it is not possible to create a secret with a specific ARN')
                            AppConfig._aws_secrets_debug_log_boto3_error(err_put_secret)

                        else:
                            Log.error('AWS Secret "%s"' % secret_id, 'does not exist - attempting to create it')
                            AppConfig._aws_secrets_debug_log_boto3_error(err_put_secret)

                            try:
                                aws_client.create_secret(Name=secret_id, ForceOverwriteReplicaSecret=False)

                            except botocore.exceptions.ClientError as err_create_secret:
                                if err_create_secret.response['Error']['Code'] == 'AccessDeniedException':
                                    Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s"' % secret_id,
                                              '- access denied: does the IAM user have the',
                                              '"secretsmanager:CreateSecret" permission?')
                                else:
                                    Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s"' % secret_id,
                                              '- attempt to create failed with unexpected error; see the proxy\'s',
                                              'debug mode log for details')
                                AppConfig._aws_secrets_debug_log_boto3_error(err_create_secret)

                            else:
                                Log.info('Created AWS Secret "%s"' % secret_id)
                                try:
                                    aws_client.put_secret_value(SecretId=secret_id,
                                                                SecretString=json.dumps(aws_secrets[secret_id]))

                                except botocore.exceptions.ClientError as err_put_secret_after_create:
                                    if err_put_secret_after_create.response['Error']['Code'] == 'AccessDeniedException':
                                        Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s"' % secret_id,
                                                  '- access denied: does the IAM user have the',
                                                  '"secretsmanager:PutSecretValue" permission?')
                                    else:
                                        # TODO: missing format string
                                        Log.error('Failed to store OAuth 2.0 tokens in AWS Secret "%s"' % secret_id,
                                                  '- unexpected error; see the proxy\'s debug mode log for details')
                                    AppConfig._aws_secrets_debug_log_boto3_error(err_put_secret_after_create)

                    elif err_put_secret.response['Error']['Code'] == 'AccessDeniedException':
                        Log.error('Failed to store OAuth tokens in AWS Secret "%s"' % secret_id, '- access denied:',
                                  'does the IAM user have the "secretsmanager:PutSecretValue" permission?')
                        AppConfig._aws_secrets_debug_log_boto3_error(err_put_secret)

                    else:
                        Log.error('Failed to store OAuth tokens in AWS Secret "%s"' % secret_id, '- unexpected error;',
                                  ' see the proxy\'s debug mode log for details')
                        AppConfig._aws_secrets_debug_log_boto3_error(err_put_secret)
        else:
            Log.error('Unable to get AWS client; cannot store OAuth 2.0 tokens in AWS Secrets Manager')

        # the updated configuration without OAuth 2.0 tokens (see AppConfig._SECRETS_MANAGER_KEYS)
        return appconfig_to_save
```

Minor edits for consistency/clarity (and fixed a few missing format strings).

Reply:
Will apply further minor changes:
- `the Secret` -> `the secret` on line 439
- Remove `# TODO: missing format string` on line 473